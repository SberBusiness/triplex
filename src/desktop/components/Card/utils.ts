import {ECardRoundingSize} from '@sberbusiness/triplex/desktop/components/Card/enums';

/** Соответствие типа скругления карточки стилевому классу. */
export const mapCardRoundingSizeToCssClass = {
    [ECardRoundingSize.MD]: 'cssClass[roundingMD]',
    [ECardRoundingSize.SM]: 'cssClass[roundingSM]',
};
