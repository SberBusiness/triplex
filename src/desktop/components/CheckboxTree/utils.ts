import {ICheckboxTreeCheckboxData} from './types';

/** Обход ICheckboxData[]. */
export const traverseCheckboxes = (checkboxes: ICheckboxTreeCheckboxData[], cb: (checkbox: ICheckboxTreeCheckboxData) => void): void => {
    checkboxes.forEach((c) => {
        if (c.children) traverseCheckboxes(c.children, cb);
        cb(c);
    });
};

/** Обновление флага checked и bulk родителя, при изменении дочернего чекбокса. */
export const checkParentCheckboxes = (checkbox: ICheckboxTreeCheckboxData): void => {
    if (!checkbox.children) {
        return;
    }

    let checkedChildrenCount = 0;
    let bulkChildrenCount = 0;

    checkbox.children.forEach((c) => {
        c.checked ? checkedChildrenCount++ : '';
        c.bulk ? bulkChildrenCount++ : '';
    });

    // Все дочерние чекбоксы выбраны.
    if (checkedChildrenCount === checkbox.children.length) {
        checkbox.checked = true;
        checkbox.bulk = bulkChildrenCount !== 0;
    } else if (checkedChildrenCount > 0) {
        // Некоторые дочерние чекбоксы выбраны.
        checkbox.checked = true;
        checkbox.bulk = true;
    } else {
        // Все дочерние чекбоксы не выбраны.
        checkbox.checked = false;
    }
};

/** Обновление флага checked дочерних чекбоксов, при изменении родителя. */
export const checkChildrenCheckboxes = (checkbox: ICheckboxTreeCheckboxData): void => {
    if (!checkbox.children) {
        return;
    }

    checkbox.children.forEach((c) => {
        c.checked = checkbox.checked;
        checkChildrenCheckboxes(c);
    });
};
