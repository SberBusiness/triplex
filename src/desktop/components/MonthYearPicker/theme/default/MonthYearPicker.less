@import '../../../../common/theme/default';

.monthYearPicker {
    position: relative;
    width: @picker-width;
}
