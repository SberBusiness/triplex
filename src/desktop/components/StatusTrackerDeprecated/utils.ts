import {
    EStatusTrackerDeprecatedStepStatus,
    StatusTrackerDeprecatedStatus,
    StatusTrackerDeprecatedStepNumber,
} from '@sberbusiness/triplex/desktop/components/StatusTrackerDeprecated/enums';
import {EStepStatus} from '@sberbusiness/triplex/desktop/components/Step/enums';

/** Мап общего статуса компонента в статус конкретного шага.  */
const trackerStatusToStepStatusMap = {
    [StatusTrackerDeprecatedStatus.WAIT]: EStatusTrackerDeprecatedStepStatus.WAIT,
    [StatusTrackerDeprecatedStatus.WARNING]: EStatusTrackerDeprecatedStepStatus.WARNING,
    [StatusTrackerDeprecatedStatus.SUCCESS]: EStatusTrackerDeprecatedStepStatus.SUCCESS,
    [StatusTrackerDeprecatedStatus.ERROR]: EStatusTrackerDeprecatedStepStatus.ERROR,
};

/** Маппинг общего статуса компонента в статус конкретного шага.  */
export const mapTrackerStatusToStepStatus = (
    currentStep: StatusTrackerDeprecatedStepNumber,
    status: StatusTrackerDeprecatedStatus,
    iteratorStep: StatusTrackerDeprecatedStepNumber
): EStatusTrackerDeprecatedStepStatus => {
    let result;

    if (currentStep > iteratorStep) {
        result = EStatusTrackerDeprecatedStepStatus.SUCCESS;
    } else if (currentStep < iteratorStep) {
        result = EStatusTrackerDeprecatedStepStatus.DISABLED;
    } else {
        result = trackerStatusToStepStatusMap[status];
        if (result === undefined) {
            throw Error(`StatusTrackerStatus not mapped: '${status}'.`);
        }
    }

    return result;
};

/**
 * Мапа статусов шагов трекера в шаги маркера.
 *
 * TODO: Выпилить в ближайшем мажорном релизе, перевести на общий энам.
 */
const statusTrackerStepStatusToStepMarkerStatusMap = {
    [EStatusTrackerDeprecatedStepStatus.WAIT]: EStepStatus.WAIT,
    [EStatusTrackerDeprecatedStepStatus.WARNING]: EStepStatus.WARNING,
    [EStatusTrackerDeprecatedStepStatus.ERROR]: EStepStatus.ERROR,
    [EStatusTrackerDeprecatedStepStatus.SUCCESS]: EStepStatus.SUCCESS,
    [EStatusTrackerDeprecatedStepStatus.DISABLED]: EStepStatus.DISABLED,
};

/**
 * Мапинг статусов шагов трекера в шаги маркера.
 *
 * TODO: Выпилить в ближайшем мажорном релизе, перевести на общий энам.
 */
export const mapStatusTrackerStepStatusToStepMarkerStatus = (status: EStatusTrackerDeprecatedStepStatus): EStepStatus => {
    const result = statusTrackerStepStatusToStepMarkerStatusMap[status];

    if (result === undefined) {
        throw Error(`EStatusTrackerStepStatus not mapped: '${status}'.`);
    }

    return result;
};
