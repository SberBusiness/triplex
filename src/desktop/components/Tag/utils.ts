import {ETagSize} from '@sberbusiness/triplex/desktop/components/Tag/enums';

/**
 * Мапа энама размера тэга на стилевой класс.
 */
export const mapTagSizeToCssClass = {
    [ETagSize.SM]: 'cssClass[sm]',
    [ETagSize.MD]: 'cssClass[md]',
};
