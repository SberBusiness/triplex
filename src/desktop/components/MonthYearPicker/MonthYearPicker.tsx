import * as React from 'react';
import moment, {Moment} from 'moment';
import {MonthYearInput} from '@sberbusiness/triplex/desktop/components/MonthYearPicker/components/MonthYearInput';
import {inputMonthYearFormat} from '@sberbusiness/triplex/desktop/components/MonthYearPicker/const';
import {Calendar} from '@sberbusiness/triplex/desktop/components/Calendar/Calendar';
import {ICalendarNestedProps} from '@sberbusiness/triplex/desktop/components/Calendar/types';
import {EPickType} from '@sberbusiness/triplex/desktop/components/Calendar/enums';
import {Dropdown, EDropdownAlignment} from '@sberbusiness/triplex/desktop/components/Dropdown/Dropdown';
import {IDateLimitRange} from '@sberbusiness/triplex/desktop/common/types/DateTypes';
import {dateFormatYYYYMMDD, globalLimitRange} from '@sberbusiness/triplex/desktop/common/consts/DateConst';
import {getFormattedDate} from '@sberbusiness/triplex/desktop/utils/dateUtils';
import {classnames} from '@sberbusiness/triplex/common/utils/classnames/classnames';

/** Свойства MonthYearPicker. */
export interface IMonthYearPickerProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange' | 'onBlur'>, ICalendarNestedProps {
    /** Значение даты текстом. */
    value: string;
    /** Callback для события изменения значения. */
    onChange: (date: string) => void;
    /** Callback для события потери фокуса. */
    onBlur?: () => void;
    /** Выравнивание календаря по горизонтали. */
    alignment?: EDropdownAlignment;
    /** Диапазон даты начала и конца периода ограничения. */
    limitRange?: IDateLimitRange;
    /** Формат для значения. */
    format?: string;
    /** Состояние ошибки. */
    error?: boolean;
    /** Отключённое состояние. */
    disabled?: boolean;
    /** Текст подсказки. */
    placeholder?: string;
    /** Обратный порядок выбора даты. */
    reversedPick?: boolean;
    /** Использование запрещено. */
    children?: never;
}

/** Состояния MonthYearPicker. */
interface IMonthYearPickerState {
    /** Value для компонента Calendar. */
    calendarValue: Moment;
    /** Фокус в поле открывает popover с выбором даты, а так же позволяет вводить дату вручную. */
    focused: boolean;
    /** Value для компонента DateInput. */
    inputValue: string;
    /** Последняя валидная дата. */
    lastValidValue: string;
    /** Диапазон даты начала и конца периода ограничения. */
    limitRange: IDateLimitRange;
}

/**
 * Проверяет дату на лимиты.
 *
 * @prop {Moment} date Дата для проверки.
 * @prop {IDateLimitRange} limitRange Диапазон даты начала и конца периода ограничения.
 */
function checkOutOfRange(date: Moment, limitRange: IDateLimitRange): boolean {
    return date.isBefore(limitRange.dateFrom) || date.isAfter(limitRange.dateTo);
}

/** Компонент ввода и выбора даты. */
export class MonthYearPicker extends React.PureComponent<IMonthYearPickerProps, IMonthYearPickerState> {
    public static defaultProps = {
        format: dateFormatYYYYMMDD,
        limitRange: globalLimitRange,
        alignment: EDropdownAlignment.LEFT,
    };

    private readonly targetRef: React.RefObject<HTMLDivElement>;
    private readonly dropdownRef: React.RefObject<HTMLDivElement>;

    constructor(props: IMonthYearPickerProps) {
        super(props);
        const {value, format, limitRange} = props;

        this.state = {
            calendarValue: this.getCalendarValueFromValue(),
            focused: false,
            lastValidValue: moment(value, format, true).isValid() ? value : '',
            inputValue: this.getInputValueFromValue(),
            limitRange: {
                dateFrom: limitRange?.dateFrom || globalLimitRange.dateFrom,
                dateTo: limitRange?.dateTo || globalLimitRange.dateTo,
            },
        };

        this.targetRef = React.createRef();
        this.dropdownRef = React.createRef();
    }

    public componentDidMount(): void {
        document.addEventListener('mousedown', this.handleClickOutside);
        document.addEventListener('touchstart', this.handleClickOutside);
    }

    public componentDidUpdate(prevProps: Readonly<IMonthYearPickerProps>): void {
        const {value, format, limitRange} = this.props;
        const {limitRange: stateLimitRange} = this.state;
        if (value != prevProps.value) {
            this.setLastValidValue(value, format);
            this.setCalendarValue();
            this.setInputValue();
        }

        const dateFrom = limitRange?.dateFrom || globalLimitRange.dateFrom;
        const dateTo = limitRange?.dateTo || globalLimitRange.dateTo;

        if (!stateLimitRange.dateFrom?.isSame(dateFrom) || !stateLimitRange.dateTo?.isSame(dateTo)) {
            this.setState({
                limitRange: {
                    dateFrom,
                    dateTo,
                },
            });
        }
    }

    public componentWillUnmount(): void {
        document.removeEventListener('mousedown', this.handleClickOutside);
        document.removeEventListener('touchstart', this.handleClickOutside);
    }

    public render(): JSX.Element {
        const {
            className,
            value,
            format,
            error,
            disabled,
            alignment,
            limitRange: limitRangeProp,
            onChange,
            onBlur,
            reversedPick,
            dayHtmlAttributes,
            monthHtmlAttributes,
            yearHtmlAttributes,
            prevButtonProps,
            nextButtonProps,
            changeViewLinkProps,
            placeholder,
            ...htmlProps
        } = this.props;
        const {calendarValue, focused, inputValue, limitRange} = this.state;

        const classNames = classnames('cssClass[monthYearPicker]', className);

        return (
            <div className={classNames} ref={this.targetRef}>
                <MonthYearInput
                    {...htmlProps}
                    value={inputValue}
                    focused={focused}
                    disabled={disabled}
                    error={error}
                    onFocus={this.handleFocus}
                    placeholder={placeholder}
                />
                <Dropdown opened={focused} targetRef={this.targetRef} forwardedRef={this.dropdownRef} alignment={alignment}>
                    <Calendar
                        pickedDate={calendarValue}
                        limitRange={limitRange}
                        onChangeDate={this.handleValueChangeFromCalendar}
                        onBlur={this.handleBlur}
                        pickType={EPickType.monthYearPick}
                        reversedPick={reversedPick}
                        dayHtmlAttributes={dayHtmlAttributes}
                        monthHtmlAttributes={monthHtmlAttributes}
                        yearHtmlAttributes={yearHtmlAttributes}
                        prevButtonProps={prevButtonProps}
                        nextButtonProps={nextButtonProps}
                        changeViewLinkProps={changeViewLinkProps}
                    />
                </Dropdown>
            </div>
        );
    }

    private setInputValue = (): void => {
        const {inputValue} = this.state;
        const nextInputValue = this.getInputValueFromValue();

        if (inputValue !== nextInputValue) {
            this.setState({
                inputValue: nextInputValue,
            });
        }
    };

    private getInputValueFromValue = (): string => {
        const {format, value} = this.props;
        const date = moment(value, format, true);

        let inputValue = value;

        if (value && date.isValid()) {
            inputValue = date.format(inputMonthYearFormat);
        }

        return inputValue;
    };

    private setCalendarValue = (): void => {
        const {calendarValue} = this.state;
        const nextCalendarValue = this.getCalendarValueFromValue();

        if (!moment(nextCalendarValue).isSame(calendarValue)) {
            this.setState({
                calendarValue: nextCalendarValue,
            });
        }
    };

    private getCalendarValueFromValue = (): moment.Moment => {
        const {format, value} = this.props;
        const lastValidValue = this.state ? this.state.lastValidValue : '';
        const date = moment(value, format, true);
        let calendarValue = moment();

        if (value) {
            if (date.isValid()) {
                calendarValue = date;
            } else if (lastValidValue) {
                calendarValue = moment(lastValidValue);
            }
        }

        return calendarValue;
    };

    /** Обработчик установки фокуса. */
    private handleFocus = (): void => {
        const {disabled} = this.props;
        if (!disabled) {
            this.setState({focused: true});
        }
    };

    /** Обработчик потери фокуса. */
    private handleBlur = (): void => {
        this.setState({focused: false});

        // Необходимо чтобы onBlur был всегда позже onChange (Не правильная последовательность происходит
        // при замене не валидного значения валидным при потере фокуса)
        setTimeout(() => {
            if (this.props.onBlur) {
                this.props.onBlur();
            }
        });
    };

    /**
     * Обработчик изменения даты из календаря.
     * @param {Moment} date Новое значение поля.
     */
    private handleValueChangeFromCalendar = (date: Moment): void => {
        const {onChange, format} = this.props;

        onChange(getFormattedDate(date, format));

        this.handleBlur();
    };

    /** При клике вне компонента скрываем выпадающий список. */
    private handleClickOutside = (event: UIEvent) => {
        const {focused} = this.state;
        const {current: target} = this.targetRef;
        const {current: dropdown} = this.dropdownRef;

        if (focused) {
            if (!target?.contains(event.target as Node) && !dropdown?.contains(event.target as Node)) {
                this.handleBlur();
            }
        }
    };

    /** Заполнение валидного значения даты. */
    private setLastValidValue = (value: string, format: string | undefined): void => {
        const {lastValidValue, limitRange} = this.state;

        const wasChanged = lastValidValue !== value;
        if (!wasChanged) {
            return;
        }

        if (!value) {
            this.setState({lastValidValue: ''});
            return;
        }

        const date = moment(value, format, true);
        const isValid = date.isValid();
        if (!isValid) {
            return;
        }

        const isOutOfRangeValue = checkOutOfRange(date, limitRange);
        if (isOutOfRangeValue) {
            return;
        }

        this.setState({lastValidValue: value});
    };
}
