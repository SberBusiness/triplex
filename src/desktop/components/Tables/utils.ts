import {EScreenWidth} from '@sberbusiness/triplex/common/enums/EScreenWidth';
import {ECellType, EHorizontalAlign, EVerticalAlign} from '@sberbusiness/triplex/desktop/components/Tables/TableBasic/enums';
import {cssClass} from '@sberbusiness/triplex/desktop/utils/cssClass';

/**
 * Преобразование значения горизонтального выравнивания в css-класс.
 * @param {EHorizontalAlign} [horizontalAlign] Значение для маппинга.
 */
export const mapHorizontalAlignToClassName = (horizontalAlign: EHorizontalAlign = EHorizontalAlign.LEFT): string => {
    switch (horizontalAlign) {
        case EHorizontalAlign.LEFT:
            return 'cssClass[alignLeft]';
        case EHorizontalAlign.RIGHT:
            return 'cssClass[alignRight]';
        case EHorizontalAlign.CENTER:
            return 'cssClass[alignCenter]';
    }
};

/**
 * Преобразование значения вертикального выравнивания в css-класс.
 * @param {EVerticalAlign} [verticalAlign] Значение для маппинга.
 */
export const mapVerticalAlignToClassName = (verticalAlign: EVerticalAlign = EVerticalAlign.BASELINE): string => {
    switch (verticalAlign) {
        case EVerticalAlign.BASELINE:
            return 'cssClass[verticalAlignBaseline]';
        case EVerticalAlign.SUB:
            return 'cssClass[verticalAlignSub]';
        case EVerticalAlign.SUPER:
            return 'cssClass[verticalAlignSuper]';
        case EVerticalAlign.TEXT_TOP:
            return 'cssClass[verticalAlignTextTop]';
        case EVerticalAlign.TEXT_BOTTOM:
            return 'cssClass[verticalAlignTextBottom]';
        case EVerticalAlign.MIDDLE:
            return 'cssClass[verticalAlignMiddle]';
        case EVerticalAlign.TOP:
            return 'cssClass[verticalAlignTop]';
        case EVerticalAlign.BOTTOM:
            return 'cssClass[verticalAlignBottom]';
    }
};

/**
 * Преобразование значений скрытия колонок в css-классы.
 * @param {EScreenWidth} [size] Массив размеров ширины экрана, при которых скрывать колонку.
 */
export const mapScreenSizeToClassName = (size?: EScreenWidth): string | undefined => {
    switch (size) {
        case EScreenWidth.XS_MAX:
            return 'cssClass[hideXS]';
        case EScreenWidth.SM_MAX:
            return 'cssClass[hideSM]';
        case EScreenWidth.MD_MAX:
            return 'cssClass[hideMD]';
        case EScreenWidth.LG_MAX:
            return 'cssClass[hideLG]';
    }
};

/**
 * Преобразование типа ячейки в css-класс.
 * @param {ECellType} [cellType] Тип ячейки.
 */
export const mapCellTypeToClassName = (cellType: ECellType = ECellType.TEXT): string => cssClass(`${cellType}Type`);
