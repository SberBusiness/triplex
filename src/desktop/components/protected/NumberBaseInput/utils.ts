import {formatAmount} from '@sberbusiness/triplex/desktop/utils/amountUtils';
import {AmountConst} from '@sberbusiness/triplex/desktop/common/consts/AmountConst';

/**
 * Вычисляем позицию каретки.
 *
 * @param {string} prevValueFormatted Прошлое отформатированное значение.
 * @param {string} newValueRaw Новое не отформатированное значение.
 * @param {number} fractionLength Количество знаков в дробной части. При значении 0 подразумевается целочисленное значение.
 * @param {number} currentCaretPosition Позиция каретки в поле ввода.
 */
export const countCaret = (prevValueFormatted: string, newValueRaw: string, fractionLength: number, currentCaretPosition: number) => {
    const spacesInOldValue = (prevValueFormatted.match(/\s/g) || []).length;
    const spacesInNewValue = (formatAmount(newValueRaw, fractionLength).match(/\s/g) || []).length;
    let caretSpaceOffset = spacesInNewValue - spacesInOldValue;

    if (prevValueFormatted.length - newValueRaw.length > 1) {
        // Если было удалено больше одного символа. Делаем reset caretSpaceOffset.
        caretSpaceOffset = 0;
    } else if (newValueRaw.substr(0, 1) === AmountConst.GroupSeparator) {
        // Если впереди пробел (" 786 673") сдвигаем каретку.
        caretSpaceOffset++;
    }

    return currentCaretPosition + caretSpaceOffset;
};
