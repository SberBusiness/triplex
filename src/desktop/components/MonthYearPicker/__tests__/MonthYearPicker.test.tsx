import * as React from 'react';
import moment from 'moment';
import renderer from 'react-test-renderer';
import {getFormattedDate} from '@sberbusiness/triplex/desktop/utils/dateUtils';
import {MonthYearPicker} from '../MonthYearPicker';

jest.mock('@sberbusiness/icons/CalendarSrvIcon20', () => ({
    CalendarSrvIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/CaretleftSrvxIcon24', () => ({
    CaretleftSrvxIcon24: 'svg',
}));

jest.mock('@sberbusiness/icons/CaretrightSrvxIcon24', () => ({
    CaretrightSrvxIcon24: 'svg',
}));

describe('MonthYearPicker', () => {
    (global as any).document = {
        addEventListener: jest.fn(),
    };

    it('renders correctly', () => {
        const handleChange = jest.fn();
        const value = getFormattedDate(moment('фев. 2021'));
        const tree = renderer.create(<MonthYearPicker value={value} onChange={handleChange} />).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
