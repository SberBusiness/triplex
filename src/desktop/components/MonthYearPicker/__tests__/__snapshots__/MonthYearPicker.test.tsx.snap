// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthYearPicker renders correctly 1`] = `
<div
  className="cssClass[monthYearPicker]"
>
  <div
    className="cssClass[inputBlock]"
  >
    <div
      className="cssClass[pseudoInput]"
      onFocus={[Function]}
      tabIndex={0}
    >
      Jan 2021
    </div>
    <span
      aria-hidden="true"
      className="cssClass[calendarIcon]"
      onMouseDown={[Function]}
    >
      <svg />
    </span>
  </div>
</div>
`;
